Dockerfile.fedora

```diff
@@ -13,3 +15,15 @@ COPY . /jcoz
 WORKDIR /jcoz
 
 RUN make -j`nproc` all
+
+FROM fedora:${FEDORA_VERSION}
+
+RUN yum install -y java-1.8.0-openjdk-devel
```

Comment:
Are these lines required? same for WORKDIR /jcoz below

Reply:
They are not _required_, but this turns the process into a multi-stage build which means that we dont bother including the layers used to build jcoz - and only take the artifacts from the build along with openjdk which is needed to run it. It's what we would want to push to a Docker registry if we were going to publish images. 

Quick example:

```
mark@carbon:~/github/JCoz.master$ docker build . -f Dockerfile.fedora -t jcoz_fedora
Sending build context to Docker daemon  9.164MB
Step 1/13 : ARG FEDORA_VERSION=30
Step 2/13 : FROM fedora:${FEDORA_VERSION} as build
 ---> 9eff96f4c827
Step 3/13 : RUN yum install -y     gcc-c++     java-1.8.0-openjdk-devel     make     maven     spdlog-devel     which
 ---> Using cache
 ---> c2d22ea2d9d3
Step 4/13 : COPY . /jcoz
 ---> Using cache
 ---> 6db57391c529
Step 5/13 : WORKDIR /jcoz
 ---> Using cache
 ---> e28330a0176b
Step 6/13 : RUN make -j`nproc` all
 ---> Using cache
 ---> 0050bd9d411f
Step 7/13 : FROM fedora:${FEDORA_VERSION}
 ---> 9eff96f4c827
Step 8/13 : RUN yum install -y java-1.8.0-openjdk-devel
 ---> Using cache
 ---> f9388ac8d1e2
Step 9/13 : COPY --from=build /jcoz/build*/liblagent.so /jcoz/
 ---> Using cache
 ---> 11d6e9ebd358
Step 10/13 : COPY --from=build /jcoz/src/java/target/*.jar /jcoz/
 ---> Using cache
 ---> 0dc22bbe3b14
Step 11/13 : COPY --from=build /jcoz/src/java/src/test/java/test /jcoz/test/
 ---> Using cache
 ---> dde686e9cf62
Step 12/13 : WORKDIR /jcoz
 ---> Using cache
 ---> 62a0a55dfd7a
Step 13/13 : CMD ["/bin/bash"]
 ---> Using cache
 ---> 1adff8f59236
Successfully built 1adff8f59236
Successfully tagged jcoz_fedora:latest
```
0050bd9d411f is the resulting image after we do the `make all`

```
mark@carbon:~/github/JCoz.master$ docker image ls | grep 0050bd9d411f
<none>              <none>              0050bd9d411f        6 minutes ago       964MB
```
1adff8f59236 aka jcoz_fedora:latest is the end result:
```
mark@carbon:~/github/JCoz.master$ docker image ls jcoz_fedora:latest
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
jcoz_fedora         latest              1adff8f59236        5 minutes ago       722MB

```
Here we can see the resulting image is 722MB versus 964MB... the image is still huge compared to the Debian one.

If you check the history (oldest layer at the bottom), you can see that most of the size (465MB) is due to the installation of openjdk:

```
mark@carbon:~/github/JCoz.master$ docker history jcoz_fedora:latest
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
1adff8f59236        7 minutes ago       /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B                  
62a0a55dfd7a        7 minutes ago       /bin/sh -c #(nop) WORKDIR /jcoz                 0B                  
dde686e9cf62        7 minutes ago       /bin/sh -c #(nop) COPY dir:4748ebcef01f5c430…   9.18kB              
0dc22bbe3b14        7 minutes ago       /bin/sh -c #(nop) COPY multi:85df31b1b5052b8…   179kB               
11d6e9ebd358        7 minutes ago       /bin/sh -c #(nop) COPY file:6ec8a484d73991ca…   9.04MB              
f9388ac8d1e2        3 days ago          /bin/sh -c yum install -y java-1.8.0-openjdk…   465MB               
9eff96f4c827        12 days ago         /bin/sh -c #(nop)  CMD ["/bin/bash"]            0B                  
<missing>           12 days ago         /bin/sh -c #(nop) ADD file:e73a501dd61569485…   248MB               
<missing>           12 days ago         /bin/sh -c #(nop)  ENV DISTTAG=f30-updates-c…   0B                  
<missing>           8 months ago        /bin/sh -c #(nop)  LABEL maintainer=Clement …   0B
```

Reply:
Got it, thanks a lot for the detailed explanation and for educating a container noob on these kinds of things. Your change looks great - would you mind at some point adding a header comment to these Dockerfiles that describe the build process? Perhaps we should add some details to the wiki as well?

In the meantime, LGTM!